sys/hoon.hoon

```diff
@@ -11598,23 +11397,3309 @@
+::    recreate this structure in the printing code, and that's what we're
+::    doing now.
+::
+::  - XX The xpat of an xray could be computed on demand instead of
+::    up-front. Possibly a lot faster!
+::
+::  - XX The loop-detection of an xray could be done on demand instead
+::    of up-front. Possibly a lot faster!
+::
+::  - XX The xpat matching code is basically brute-force.
+::
+::    If it turns out to be a performance bottleneck, there's lots of
+::    low-hanging fruit there. For example:
+::
+::    - Faces repeat the work done for the type they reference.
+::    - When detecting whether a cell is part of an "informal" list,
+::      we recurse into the tail repeatedly. For example, the following
+::      example will do the "formal-list" test 3 times:
+::
+::      - `[1 2 `(list @)`~]`
+::
+::  - XX Try to find a way to drop the `%pntr` constructor from
+::    `%xdat`. The consumer of an `xray` does not care.
+::
+::  - XX Actually, it would also be really nice to produce another
+::    version of this structure that doesn't have the (unit *) wrapper around
+::    everything interesting. This would make the on-demand computation
+::    of various things hard, though.
+::
+::  - XX Simply lying about the type of deep arms is not robust. I am just
+::    claiming that they are nouns, but if another thing in the xray
+::    actually needs it, it will think it's a noun too.
+::
+::  - XX There are probably remaining bugs. Test the shit out of this.
+::
+::  - XX What should the `xrole` of a cell with a %noun head be? I
+::    think the current design can't handle this case coherently.
+::
+++  json                                                ::  normal json value
+  $@  ~                                                 ::  null
+  $%  {$a p/(list json)}                                ::  array
+      {$b p/?}                                          ::  boolean
+      {$o p/(map @t json)}                              ::  object
+      {$n p/@ta}                                        ::  number
+      {$s p/@t}                                         ::  string
+  ==                                                    ::
+::
+++  libxray
+  ::
+  |^  ^-  $:  ximage-to-spec=$-(=ximage =spec)
+              xray-type=$-([@ type] ximage)
+              focus-on=$-([xtable xkey] xray)
+          ==
+      [ximage-to-spec xray-type focus-on]
+  ::
+  +|  %utils
+  ::
+  ::  Left-fold over a list.
+  ::
+  ++  fold
+     |*  [state=mold elem=mold]
+     |=  [[st=state xs=(list elem)] f=$-([state elem] state)]
+     ^-  state
+     |-
+     ?~  xs  st
+     =.  st  (f st i.xs)
+     $(xs t.xs, st st)
+  ::
+  ::  This is basically a `mapM` over a list using the State monad.
+  ::
+  ::  Another way to think about this is that it is the same as `turn`,
+  ::  except that a state variable `st` is threaded through the
+  ::  execution. The list is processed from left to right.
+  ::
+  ++  traverse
+    |*  [state=mold in=mold out=mold]
```

Comment:
`in` shadows the set core, which we should probably not do in general. +traverse-set uses +input, so I think this should switch to that

Reply:
I think these should really be `gate`, not `mold`, unless there's some variance issue.

Reply:
Oh, nevermind, I see why they're molds.